File: Core/Action.swift
import AppKit
import CoreGraphics
import os

/// A single thing the controller can do when a button is pressed.
///
/// `Action` is a simple value type that wraps a closure. All the complexity
/// of *how* to do something lives in the executor layer (`KeystrokeExecutor`,
/// `AppleScriptExecutor`). Modes just declare *what* they want, using the
/// static factory methods below.
///
/// The `id` field is stable across launches and is used for persisting custom
/// button mappings to disk. The `displayName` is shown in the preferences UI.
struct Action {

    /// A stable, URL-safe identifier. Use kebab-case (e.g. `"spotify-next-track"`).
    let id: String

    /// A human-readable description shown in the preferences window.
    let displayName: String

    /// The work to perform. Threading is the factory method's responsibility:
    /// keystroke actions fire synchronously, AppleScript actions dispatch async.
    let perform: () -> Void
}

// MARK: - Factory Methods

extension Action {

    /// Sends a keyboard shortcut to the currently active application.
    ///
    /// This fires instantly and requires the Accessibility permission granted
    /// at first launch.
    static func keystroke(
        _ key: KeyCode,
        modifiers: CGEventFlags = [],
        id: String,
        name: String
    ) -> Action {
        Action(id: id, displayName: name) {
            KeystrokeExecutor.send(key: key, modifiers: modifiers)
        }
    }

    /// Runs an AppleScript source string on a background queue.
    ///
    /// AppleScript execution can block for several hundred milliseconds, so it
    /// is always dispatched off the main thread to keep the UI responsive.
    static func appleScript(
        _ source: String,
        id: String,
        name: String
    ) -> Action {
        Action(id: id, displayName: name) {
            AppleScriptExecutor.run(source)
        }
    }

    /// Brings an application to the front, launching it if it isn't running.
    static func openApp(
        bundleIdentifier: String,
        id: String,
        name: String
    ) -> Action {
        Action(id: id, displayName: name) {
            guard let url = NSWorkspace.shared.urlForApplication(
                withBundleIdentifier: bundleIdentifier
            ) else {
                Logger.actions.error("Cannot find application '\(bundleIdentifier)'")
                return
            }
            let config = NSWorkspace.OpenConfiguration()
            config.activates = true
            NSWorkspace.shared.openApplication(
                at: url,
                configuration: config,
                completionHandler: nil
            )
        }
    }

    /// Changes the system output volume by a relative number of points (0–100 scale).
    ///
    /// Positive values raise the volume; negative values lower it.
    /// Clamped to [0, 100] by the system.
    static func adjustVolume(
        by delta: Int,
        id: String,
        name: String
    ) -> Action {
        Action(id: id, displayName: name) {
            // AppleScript is used here because there is no clean public API
            // for system volume adjustment without private frameworks.
            AppleScriptExecutor.run("""
                set currentVolume to output volume of (get volume settings)
                set volume output volume (currentVolume + \(delta))
            """)
        }
    }

    /// Toggles the system mute state.
    static func toggleMute(id: String, name: String) -> Action {
        Action(id: id, displayName: name) {
            AppleScriptExecutor.run("""
                set isMuted to output muted of (get volume settings)
                set volume output muted (not isMuted)
            """)
        }
    }

    /// Sends a macOS screenshot command.
    static func screenshot(region: Bool, id: String, name: String) -> Action {
        // Cmd+Shift+4 = region select; Cmd+Shift+3 = full screen
        let key: KeyCode = region ? .four : .three
        return keystroke(
            key,
            modifiers: [.maskCommand, .maskShift],
            id: id,
            name: name
        )
    }

    /// Switches to the previous macOS Space (virtual desktop).
    static func switchSpace(direction: HorizontalDirection, id: String, name: String) -> Action {
        let key: KeyCode = direction == .left ? .leftArrow : .rightArrow
        return keystroke(key, modifiers: .maskControl, id: id, name: name)
    }

    /// A no-op placeholder for unassigned buttons.
    ///
    /// Useful during development and as a default for new buttons added in
    /// future controller hardware revisions.
    static func unassigned(button: GamepadButton) -> Action {
        Action(id: "unassigned-\(button.rawValue.lowercased())", displayName: "Unassigned") {}
    }

    // MARK: - Helpers

    enum HorizontalDirection {
        case left, right
    }
}

// MARK: - KeyCode Extension for Screenshot

private extension KeyCode {
    static let three: KeyCode = .ansi3
    static let four: KeyCode = .ansi4
}

File: buttonbridge/core/action.py
"""
Action — the single thing a button press can do.

Actions are plain dataclasses wrapping a callable.  All the complexity of
*how* to do something lives in the executor layer.  Modes just declare
*what* they want using the factory methods at the bottom of this file.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Callable

from ..constants import BundleID, Spotify, Volume

logger = logging.getLogger(__name__)


@dataclass
class Action:
    """
    A single thing the controller can do when a button is pressed.

    Attributes:
        id:           Stable kebab-case identifier (e.g. ``"spotify-next-track"``).
                      Used for serialising custom mappings to disk.
        display_name: Human-readable label shown in the preferences UI.
        perform:      Zero-argument callable that carries out the action.
                      Threading is the factory method's responsibility.
    """

    id: str
    display_name: str
    perform: Callable[[], None] = field(repr=False)

    def __call__(self) -> None:
        """Convenience: ``action()`` is equivalent to ``action.perform()``."""
        logger.debug("Action fired: %s", self.id)
        self.perform()


# ---------------------------------------------------------------------------
# Factory methods
#
# Import executors inside each factory to avoid circular imports and to keep
# the top-level module light.  The callables are captured in closures so the
# import only happens once at definition time.
# ---------------------------------------------------------------------------

def keystroke(
    key: str,
    modifiers: list[str] | None = None,
    *,
    id: str,
    name: str,
) -> Action:
    """
    Send a keyboard shortcut to the active application.

    Args:
        key:       Key name as a string (e.g. ``"t"``, ``"left_arrow"``).
                   Must match a key defined in ``KeystrokeExecutor.Key``.
        modifiers: List of modifier names (``"cmd"``, ``"shift"``,
                   ``"opt"``, ``"ctrl"``).  Defaults to no modifiers.
        id:        Stable action identifier.
        name:      Display name for the preferences UI.
    """
    from ..executors.keystroke_executor import KeystrokeExecutor

    _modifiers = modifiers or []

    return Action(
        id=id,
        display_name=name,
        perform=lambda: KeystrokeExecutor.send(key, _modifiers),
    )


def apple_script(
    source: str,
    *,
    id: str,
    name: str,
) -> Action:
    """
    Run an AppleScript source string (dispatched to a background thread).

    Args:
        source: Valid AppleScript source.
        id:     Stable action identifier.
        name:   Display name.
    """
    from ..executors.applescript_executor import AppleScriptExecutor

    return Action(
        id=id,
        display_name=name,
        perform=lambda: AppleScriptExecutor.run(source),
    )


def open_app(
    bundle_identifier: str,
    *,
    id: str,
    name: str,
) -> Action:
    """
    Bring an application to the front, launching it if not already running.

    Args:
        bundle_identifier: CFBundleIdentifier of the target app.
        id:                Stable action identifier.
        name:              Display name.
    """
    from AppKit import NSWorkspace  # type: ignore[import]

    def _open() -> None:
        url = NSWorkspace.sharedWorkspace().URLForApplicationWithBundleIdentifier_(
            bundle_identifier
        )
        if url is None:
            logger.error("Cannot find application with bundle ID: %s", bundle_identifier)
            return
        config = NSWorkspace.sharedWorkspace()
        NSWorkspace.sharedWorkspace().openApplicationAtURL_configuration_completionHandler_(
            url, None, None
        )

    return Action(id=id, display_name=name, perform=_open)


def adjust_volume(delta: int, *, id: str, name: str) -> Action:
    """
    Raise or lower the system output volume by ``delta`` points (0–100 scale).

    Positive values raise the volume; negative values lower it.
    """
    from ..executors.applescript_executor import AppleScriptExecutor

    script = f"""
        set currentVolume to output volume of (get volume settings)
        set volume output volume (currentVolume + {delta})
    """
    return Action(id=id, display_name=name, perform=lambda: AppleScriptExecutor.run(script))


def toggle_mute(*, id: str, name: str) -> Action:
    """Toggle the system audio mute state."""
    from ..executors.applescript_executor import AppleScriptExecutor

    script = """
        set isMuted to output muted of (get volume settings)
        set volume output muted (not isMuted)
    """
    return Action(id=id, display_name=name, perform=lambda: AppleScriptExecutor.run(script))


def screenshot(region: bool, *, id: str, name: str) -> Action:
    """
    Trigger a macOS screenshot.

    Args:
        region: If ``True``, capture a region (Cmd+Shift+4).
                If ``False``, capture full screen (Cmd+Shift+3).
    """
    key = "4" if region else "3"
    return keystroke(key, ["cmd", "shift"], id=id, name=name)


def switch_space(direction: str, *, id: str, name: str) -> Action:
    """
    Switch to the macOS Space (virtual desktop) to the left or right.

    Args:
        direction: ``"left"`` or ``"right"``.
    """
    key = "left_arrow" if direction == "left" else "right_arrow"
    return keystroke(key, ["ctrl"], id=id, name=name)


def unassigned(button_name: str) -> Action:
    """Placeholder for intentionally unassigned buttons."""
    return Action(
        id=f"unassigned-{button_name.lower().replace(' ', '-')}",
        display_name="Unassigned",
        perform=lambda: None,
    )
